freud/diffraction.pyx

```diff
@@ -0,0 +1,314 @@
+                View orientation. Uses :math:`(1, 0, 0, 0)` if not provided
+                or :code:`None` (Default value = :code:`None`).
+            zoom (float):
+                Scaling factor for incident wavevectors (Default value = 4).
+            peak_width (float):
+                Width of Gaussian convolved with points, in system length units
+                (Default value = 1).
+        """
+        system = freud.locality.NeighborQuery.from_system(system)
+
+        if view_orientation is None:
+            view_orientation = np.array([1., 0., 0., 0.])
+
+        grid_size = int(self.grid_size / zoom)
+
+        # Compute the box projection matrix
+        inv_shear = self._calc_proj(view_orientation, system.box)
+
+        # Rotate points by the view quaternion and shear by the box projection
+        xy = rowan.rotate(view_orientation, system.points)[:, 0:2]
+        xy = xy @ inv_shear.T
+
+        # Map positions to [0, 1] and compute a histogram "image"
+        xy += 0.5
+        xy %= 1
+        im, _, _ = np.histogram2d(
+            xy[:, 0], xy[:, 1], bins=np.linspace(0, 1, grid_size))
+
+        # Compute FFT and convolve with Gaussian
+        cdef double complex[:, :] diffraction_fft
+        diffraction_fft = np.fft.fft2(im)
+        diffraction_fft = scipy.ndimage.fourier.fourier_gaussian(
+            diffraction_fft, peak_width / zoom)
+        diffraction_fft = np.fft.fftshift(diffraction_fft)
+
+        # Compute the squared modulus of the FFT, which is S(q)
+        self._diffraction = np.real(
+            diffraction_fft * np.conjugate(diffraction_fft))
+
+        # Transform the image (scale, shear, zoom) and normalize S(q) by N^2
+        N = len(system.points)
+        self._diffraction = self._transform(
+            self._diffraction, system.box, inv_shear, zoom) / (N*N)
+
+        # Compute a cached array of k-vectors that can be rotated and scaled
+        if not self._called_compute:
+            # Create a 1D axis of k-vector magnitudes
+            self._k_values_orig = np.fft.fftshift(np.fft.fftfreq(
+                n=self.output_size,
+                d=1/self.output_size))
+
+            # Create a 3D meshgrid of k-vectors with shape
+            # (output_size, output_size, 3)
+            self._k_vectors_orig = np.asarray(np.meshgrid(
+                self._k_values_orig, self._k_values_orig, [0])).T[0]
+
+        # Compute the rotated and scaled k-values and k-vectors
+        self._k_values = self._k_values_orig / np.max(system.box.to_matrix())
+        self._k_vectors = rowan.rotate(view_orientation, self._k_vectors_orig)
+        self._k_vectors /= np.max(system.box.to_matrix())
+
+        return self
+
+    @_Compute._computed_property
+    def diffraction(self):
+        """
+        (``output_size``, ``output_size``) :class:`numpy.ndarray`:
+            diffraction pattern.
+        """
+        return self._diffraction
```

Comment:
We want to be returning numpy arrays here and for the other properties, but as I've written the unit tests, I've noticed that this is not a numpy array. It is a `view.MemoryView` object. Normally to get the numpy arrays from lower level array constructs, we use `freud.util.make_managed_numpy_array`, but since this is not something taken from cpp, we may need a different way to convert it. I have used `np.asarray()` to convert in the unit tests, if that makes a copy of the underlying array data, then I think that is the best way to do the conversion.

Reply:
Good catch, @tommy-waltmann. [According to the Cython docs](https://cython.readthedocs.io/en/latest/src/userguide/memoryviews.html#coercion-to-numpy), coercion of the Cython memory view into a NumPy array with `np.asarray` should not require a copy. Can you make this change for all the properties where we return array-like quantities? Thanks!
```suggestion
        return np.asarray(self._diffraction)
```